### PR TITLE
Reimplement all of the github reporters (status, review, comment)

### PR DESCRIPTION
**Pull request by kszucs**
After a failing attempt to use buildbot's data api to store and retrieve build (benchmark) results, and another one based on the official example I had to reimplement all of the github reporters.

This PR does:
- save the output of archery benchmark command as a log named `result` for easier retrieval
- make build data retrieval possible by reimplementing GitHubReporterBase
- add a comment formatter for benchmark output (as preformatted diff - markdown tables on github don't support any kind of formatting)
- implement unit tests for reporters and formatters
- set missing timeout to 2 minutes for docker workers

### CODE REVIEW

**Comment by fsaintjacques** [comment]
mamma mia!

**Comment by fsaintjacques** [comment]
hehe, almost.

**Comment by kszucs** [comment]
Aaaaah, inline css doesn't seem to work.

**Comment by kszucs** [comment]
```diff
====  =================================  =============================================  ===========  ===========  ============
..    suite                              benchmark                                         baseline    contender        change
====  =================================  =============================================  ===========  ===========  ============
+     arrow-csv-parser-benchmark         RegressionParseCSVEscapedBlock                 7.9357e+08   7.93801e+08   0.000290762
+     arrow-csv-parser-benchmark         RegressionChunkCSVNoNewlinesBlock              1.98264      1.98264       2.91841e-07
+     arrow-compute-aggregate-benchmark  RegressionSumKernel/32768/1                    2.47803e+10  2.48177e+10   0.00150822
+     arrow-compute-aggregate-benchmark  RegressionSumKernel/32768/50                   1.92115e+10  1.91888e+10  -0.00118195
+     arrow-compute-aggregate-benchmark  RegressionSumKernel/32768/0                    2.75622e+10  2.75741e+10   0.000430551
+     arrow-compute-aggregate-benchmark  RegressionSumKernel/32768/10                   2.19336e+10  2.19115e+10  -0.00100843
+     arrow-builder-benchmark            RegressionBuildInt64DictionaryArrayRandom      9.25363e+08  9.2487e+08   -0.000532491
+     arrow-builder-benchmark            RegressionBuildFixedSizeBinaryArray            1.53788e+09  1.53656e+09  -0.000857427
+     arrow-builder-benchmark            RegressionBuildAdaptiveIntNoNulls              6.7987e+09   6.87545e+09   0.0112882
+     arrow-builder-benchmark            RegressionBuildAdaptiveIntNoNullsScalarAppend  3.03292e+09  3.05409e+09   0.0069823
+     arrow-builder-benchmark            RegressionBuildBinaryArray                     5.83293e+08  5.82337e+08  -0.00163838
+     arrow-builder-benchmark            RegressionBuildBooleanArrayNoNulls             3.93855e+09  3.91089e+09  -0.00702443
+     arrow-builder-benchmark            RegressionBuildStringDictionaryArray           5.26941e+08  5.29121e+08   0.0041355
+     arrow-builder-benchmark            RegressionBuildChunkedBinaryArray              1.07839e+09  1.07901e+09   0.000572894
+     arrow-builder-benchmark            RegressionBuildInt64DictionaryArraySequential  9.54528e+08  9.56602e+08   0.00217311
+     arrow-builder-benchmark            RegressionBuildInt64DictionaryArraySimilar     6.03168e+08  6.00542e+08  -0.00435434
+     arrow-builder-benchmark            RegressionBuildPrimitiveArrayNoNulls           2.67537e+09  2.67744e+09   0.000771888
+     arrow-csv-converter-benchmark      RegressionFloatConversion                      4.03992e+07  4.0345e+07   -0.00134217
+     arrow-csv-converter-benchmark      RegressionInt64Conversion                      1.26278e+08  1.2623e+08   -0.000381024
+     arrow-compute-filter-benchmark     RegressionCompareArrayScalarKernel/32768/10    1.30259e+10  1.24247e+10  -0.0461551
+     arrow-compute-filter-benchmark     RegressionCompareArrayScalarKernel/32768/1     1.30637e+10  1.24789e+10  -0.0447659
+     arrow-compute-filter-benchmark     RegressionCompareArrayScalarKernel/32768/50    1.30607e+10  1.24375e+10  -0.047715
+     arrow-compute-filter-benchmark     RegressionCompareArrayScalarKernel/32768/0     1.32328e+10  1.26353e+10  -0.0451514
+     arrow-bit-util-benchmark           RegressionGenerateBits/8192                    1.50754e+08  1.50546e+08  -0.00138
+     arrow-bit-util-benchmark           RegressionCopyBitmapWithoutOffset/8192         6.63549e+10  6.63927e+10   0.000568933
+     arrow-bit-util-benchmark           RegressionFirstTimeBitmapWriter/8192           1.62352e+08  1.62357e+08   2.76051e-05
+     arrow-bit-util-benchmark           RegressionBitmapWriter/8192                    1.56639e+08  1.56685e+08   0.00028979
+     arrow-bit-util-benchmark           RegressionCopyBitmapWithOffset/8192            1.00707e+09  1.00707e+09  -3.72329e-06
+     arrow-bit-util-benchmark           RegressionGenerateBitsUnrolled/8192            2.7299e+08   2.79452e+08   0.0236723
+     arrow-bit-util-benchmark           RegressionBitmapReader/8192                    1.99303e+08  1.99308e+08   2.2999e-05
====  =================================  =============================================  ===========  ===========  ============
```

**Comment by pitrou** [comment]
I think at some point it would be good to only present those regressions which are statistically significant.
It means you'll have to find the right statistical test (and also to record all samples of the baseline, not only its statistical aggregates) ;-)

**Comment by fsaintjacques** [comment]
@pitrou, you remember the `is_significant = true` method ;) ? That was the intent. Note that archery does keep an array of samples, it just exports the median.

google benchmark compare does the following:

https://github.com/google/benchmark/blob/master/tools/gbench/report.py#L157-L207

**Comment by pitrou** [comment]
@fsaintjacques Ah, right. I'll trust to find the right method ;-)

**Comment by wesm** [comment]
Is it possible to add the unit to the benchmarks and format to be more human-intelligible? The output is not very intelligible at the moment. E.g. it would be useful to see if a benchmark say went from 2GB/s to 1GB/s

Since GitHub-flavored markdown supports tables it might be worthwhile at some point to format with that

**Comment by fsaintjacques** [comment]
I'd also say to format the metrics with a formatter (bkmg for bytes, kmb/sec for items, micro,nanos, ... for timings). I'm sure there's a python library for this. To chose the width, I'd say use the delta absolute value.

**Comment by kszucs** [comment]
If You open the comments above, there are markdown/html tables which are worse than the string stables, besides We cannot use any formatting.

**Comment by kszucs** [comment]
I'm not sure whether this formatting should be implemented by ursabot, archery might be a better place for it.

File: ursabot/tests/test_reporters.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    @ensure_deferred
+    async def test_basic(self):
+        build = await self.setupBuildResults(SUCCESS)
+
+        self._http.expect(
+            'post',
+            '/repos/buildbot/buildbot/pulls/34/reviews',
+            json={
+                'event': '',
+                'commit_id': 'd34db33fd43db33f'
```
Comment (pitrou) [inline ursabot/tests/test_reporters.py]: Where does this commit id come from?

File: ursabot/formatters.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        rows = map(' '.join, zip(diff, formatted.splitlines()))
+        table = '\n'.join(rows)
+
+        return f'```diff\n{table}\n```'
+
+    def render_success(self, build, master):
+        results = self._extract_result_logs(build)
+        try:
+            # decode jsonlines objects and render the results as markdown table
+            tables = toolz.valmap(self._render_table, results)
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ursabot/formatters.py]: Each row is a step in a [build](https://ci.ursalabs.org/#/builders/48/builds/128), and each step can have multiple logs. So every step can have a `result` log, but this is controlled by [us](https://github.com/ursa-labs/ursabot/pull/66/files#diff-c5d24c345ac8870c79e4acc6e1d47df8R375).

File: ursabot/tests/test_reporters.py (hunk 1 above)
Comment (kszucs) [reply ursabot/tests/test_reporters.py]: From [here](https://github.com/buildbot/buildbot/blob/ec38d8f7bbffa4fc792d5bb6466c336015131197/master/buildbot/test/util/reporter.py#L23)

File: ursabot/formatters.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply ursabot/formatters.py]: In practice, there should be only one per build, right? Can you add a comment explaining all this?

[the remark this answers is not in the document]
Comment (kszucs) [reply ursabot/formatters.py]: Yep

File: ursabot/reporters.py [hunk 3]
```diff
@@ -12,66 +23,280 @@
...
+        else:
+            giturl = giturlparse(repo)
+            repo_owner, repo_name = giturl.owner, giturl.repo
+
+        return dict(repo=repo, branch=branch, sha=sha, issue=issue,
+                    repo_owner=repo_owner, repo_name=repo_name)
+
+    @ensure_deferred
+    async def send(self, build):
+        # XXX: the whole method is reimplemented based on the parent
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ursabot/reporters.py]: Not yet

**Comment by pitrou** [comment]
@kszucs Feel free to merge !

**Pull request merged**